onnxruntime/core/optimizer/gelu_fusion.cc

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/gelu_fusion.h"
+#include "core/graph/graph_utils.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+static void IsInputConstant(const Graph& graph, const Node& node, std::tuple<bool, NodeArg*, NodeArg*>& ret) {
+  const auto& inputs = node.InputDefs();
+  bool found_constant = false;
+  NodeArg* gelu_non_const_input = nullptr;
+  NodeArg* gelu_const_input = nullptr;
+  for (auto& i : inputs) {
+    if (graph_utils::NodeArgIsConstant(graph, *i)) {
+      // Todo: check the constant for example be sqrt(2.0) or 1
+      found_constant = true;
+      gelu_const_input = const_cast<NodeArg*>(i);
+    } else {
+      gelu_non_const_input = const_cast<NodeArg*>(i);
+    }
+  }
+
+  ret = std::make_tuple(found_constant, gelu_non_const_input, gelu_const_input);
+}
+
+static bool HasConsumers(Graph& graph, const NodeArg* arg) {
+  for (auto& node : graph.Nodes()) {
+    auto ret = std::find(node.MutableInputDefs().begin(), node.MutableInputDefs().end(), arg);
+    if (ret != node.MutableInputDefs().end()) {
+      return true;
+    }
+
+    ret = std::find(node.MutableImplicitInputDefs().begin(), node.MutableImplicitInputDefs().end(), arg);
+    if (ret != node.MutableImplicitInputDefs().end()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+Status GeluFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  std::deque<onnxruntime::NodeIndex> removed_nodes;
+  std::deque<NodeArg*> removed_initializers;
+
+  for (auto node_index : node_topology_list) {
+    auto& node = *graph.GetNode(node_index);
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Div", {7}) ||
+        !graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders()) ||
+        node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+    std::tuple<bool, NodeArg*, NodeArg*> t;
+    IsInputConstant(graph, node, t);
+    if (!std::get<0>(t) || std::get<1>(t) == nullptr || std::get<2>(t) == nullptr) {
+      continue;
```

Comment:
Can we do the check to see if you have valid inputs to continue inside IsInputConstant and just return true/false from that if everything is valid?